### PR TITLE
terraform-aws-multi-az-subnets

### PR DESCRIPTION
**Pull request by aknysh**
## what
* `terraform-aws-multi-az-subnets` implementation

## why
* We need a module to provision private or public subnets in each Availability Zone, and then be able to find the subnet IDs by the AZ names

## test
Given the following configuration

```hcl
module "vpc" {
  source     = "git::https://github.com/cloudposse/terraform-aws-vpc.git?ref=master"
  namespace  = "${var.namespace}"
  name       = "vpc"
  stage      = "${var.stage}"
  cidr_block = "${var.cidr_block}"
}

locals {
  public_cidr_block  = "${cidrsubnet(module.vpc.vpc_cidr_block, 1, 0)}"
  private_cidr_block = "${cidrsubnet(module.vpc.vpc_cidr_block, 1, 1)}"
}

module "public_subnets" {
  source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
  namespace           = "${var.namespace}"
  stage               = "${var.stage}"
  name                = "${var.name}"
  availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id              = "${module.vpc.vpc_id}"
  cidr_block          = "${local.public_cidr_block}"
  type                = "public"
  igw_id              = "${module.vpc.igw_id}"
  nat_gateway_enabled = "true"
}

module "private_subnets" {
  source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
  namespace           = "${var.namespace}"
  stage               = "${var.stage}"
  name                = "${var.name}"
  availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id              = "${module.vpc.vpc_id}"
  cidr_block          = "${local.private_cidr_block}"
  type                = "private"
  
  # Map of AZ names to NAT Gateway IDs that was created in "public_subnets" module
  az_ngw_ids          = "${module.public_subnets.az_ngw_ids}"

  # Need to explicitly provide the count since Terraform currently can't use dynamic count on computed resources from different modules
  # https://github.com/hashicorp/terraform/issues/10857
  # https://github.com/hashicorp/terraform/issues/12125
  # https://github.com/hashicorp/terraform/issues/4149
  az_ngw_count = 3
}

output "private_az_subnet_ids" {
  value = "${module.private_subnets.az_subnet_ids}"
}

output "public_az_subnet_ids" {
  value = "${module.public_subnets.az_subnet_ids}"
}
```

the output Maps of AZ names to subnet IDs look like these

```hcl
public_az_subnet_ids = {
  us-east-1a = subnet-ea58d78e
  us-east-1b = subnet-556ee131
  us-east-1c = subnet-6f54db0b
}
private_az_subnet_ids = {
  us-east-1a = subnet-376de253
  us-east-1b = subnet-9e53dcfa
  us-east-1c = subnet-a86fe0cc
}
```

and the created subnet IDs could be found by the AZ names using `map["key"]` or [`lookup(map, key, [default])`](https://www.terraform.io/docs/configuration/interpolation.html#lookup-map-key-default-),

for example:

`public_az_subnet_ids["us-east-1a"]`

`lookup(private_az_subnet_ids, "us-east-1b")`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+module "private_subnets" {
+  source             = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
+  namespace          = "${var.namespace}"
+  stage              = "${var.stage}"
+  name               = "${var.name}"
+  availability_zones = ["us-east-1a", "us-east-1b", "us-east-1c"]
+  vpc_id             = "${module.vpc.vpc_id}"
+  cidr_block         = "${local.private_cidr_block}"
+  type               = "private"
+  ngw_id             = "${module.public_subnets.ngw_id}"
```
Comment (osterman) [inline README.md]: For HA this needs to support one ngw per az

File: private.tf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  vpc_id = "${var.vpc_id}"
+
+  tags = {
+    "Name"      = "${module.private_label.id}${var.delimiter}${element(var.availability_zones, count.index)}"
+    "Stage"     = "${module.private_label.stage}"
+    "Namespace" = "${module.private_label.namespace}"
+  }
+}
+
+resource "aws_route" "private" {
```
Comment (osterman) [inline private.tf]: This should be optional. If no `ngw_id` passed, do not create the route. E.g. if simple VPC with only RDS or elasticache, does not need a ngw. I have a current use case for this.

File: outputs.tf [hunk 3]
```diff
@@ -1,16 +1,16 @@
-output "ngw_id" {
-  value       = "${join("", aws_nat_gateway.default.*.id)}"
-  description = "NAT Gateway ID"
+output "ngw_ids" {
+  value       = "${aws_nat_gateway.public.*.id}"
```
Comment (osterman) [inline outputs.tf]: This is not useful. You need to know the nat gw by AZ.

File: variables.tf [hunk 4]
```diff
@@ -43,34 +43,35 @@ variable "max_subnets" {
...
   default     = ""
 }
 
-variable "ngw_id" {
-  description = "NAT Gateway ID which will be used as a default route in private route tables (e.g. `igw-9c26a123`)"
-  default     = ""
+variable "ngw_ids" {
+  type        = "list"
+  description = "NAT Gateway IDs which are used as default routes in private route tables when creating private subnets (e.g. [`ngw-9c26a123`, `ngw-3b45a533`])"
+  default     = []
```
Comment (osterman) [inline variables.tf]: This must be a map of AZ<->`ngw_id`

File: README.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+module "private_subnets" {
+  source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=master"
+  namespace           = "${var.namespace}"
+  stage               = "${var.stage}"
+  name                = "${var.name}"
+  vpc_id              = "${module.vpc.vpc_id}"
+  cidr_block          = "${local.private_cidr_block}"
+  type                = "private"
+  nat_gateway_enabled = "false"
+  az_ngw_ids          = {
```
Comment (osterman) [inline README.md]: Is it required to pass a partially empty map of using az_ngw_ids

**Comment by comeanother** [comment]
How can we use this module, if we want to create several subnets in each AZ?
How NGW will be selected in the following cases?
* 3 public subnets, 2 private subnets
* 2 public subnets, 3 private subnets

I think, https://github.com/cloudposse/terraform-aws-named-subnets resolves all subnets distribution tasks:
* It can be instantiated per-AZ if we need Multi-AZ
* It provides a single NGW per AZ and we are free to select which NGW we want to use
* It provides capability of running several subnets in a single AZ

On the other hand, https://github.com/cloudposse/terraform-aws-dynamic-subnets provides out-of-the-box solution for creating subnets topology based on amount of AZs you want to utilize. Only option to set  amount of subnets per AZ may be required.

IMHO  https://github.com/cloudposse/terraform-aws-named-subnets provides more flexibility than https://github.com/cloudposse/terraform-aws-multi-az-subnets and https://github.com/cloudposse/terraform-aws-dynamic-subnets together.

https://github.com/cloudposse/terraform-aws-dynamic-subnets calculates subnet CIDRs automatically and saves time.

I see no benefits from https://github.com/cloudposse/terraform-aws-multi-az-subnets

@aknysh @osterman @SweetOps any thought on this?

File: private.tf [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  count          = "${local.private_count}"
+  subnet_id      = "${element(aws_subnet.private.*.id, count.index)}"
+  route_table_id = "${element(aws_route_table.private.*.id, count.index)}"
+  depends_on     = ["aws_subnet.private", "aws_route_table.private"]
+}
+
+resource "aws_route" "default" {
+  count                  = "${local.private_route_count}"
+  route_table_id         = "${lookup(local.az_route_table_ids, element(keys(var.az_ngw_ids), count.index))}"
+  nat_gateway_id         = "${lookup(var.az_ngw_ids, element(keys(var.az_ngw_ids), count.index))}"
```
Comment (comeanother) [inline private.tf]: It doesn't work as expected.
We can't have multiple `default gateways`.

**Comment by SweetOps** [comment]
use this and you'll get multi-named-multi-az :)
```terraform
locals {
subnets_count = "${length(compact(var.subnets_names) * length(compact(var.availability_zones))}"
}

resource "aws_subnet" "private" {
count             = "${local.subnets_count}"
}
```

@comeanother , mb that module need in some project... IMO we should think about implementation  functional of `named-subnets` in `dynamic-subnets`.

**Comment by comeanother** [comment]
@SweetOps It appears to me that now `dynamic-subnets` can be entierly replaced with `named-subnets`.
But for now only adding routes to particular subnets may be an issue.

**Comment by aknysh** [comment]
@comeanother  you reviewed an outdated PR, I was testing the code in my test solution and did not update this PR. Now it has the latest code and it "does work", with the exception of one thing related to multiple module invocation (see Note below).

I agree with you that `terraform-aws-named-subnets` could be used in many cases and it's more flexible and many use-cases could be covered by it. We have too many subnet modules now (three), maybe it's a good idea to write down all use-cases and combine them into one module (if possible).

__NOTE__: current issue with `terraform-aws-multi-az-subnets` invocation.
Terraform does not want to accept an input map for this module as an output from another module.
For example, this code 

```hcl
locals {
  public_cidr_block  = "${cidrsubnet(var.vpc_cidr, 1, 0)}"
  private_cidr_block = "${cidrsubnet(var.vpc_cidr, 1, 1)}"
}

module "public_subnets" {
  source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=init"
  namespace           = "${var.namespace}"
  stage               = "${var.stage}"
  name                = "${var.name}"
  availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id              = "${var.vpc_id}"
  cidr_block          = "${local.public_cidr_block}"
  type                = "public"
  igw_id              = "${var.igw_id}"
  nat_gateway_enabled = "true"
}

module "private_subnets" {
  source             = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=init"
  namespace          = "${var.namespace}"
  stage              = "${var.stage}"
  name               = "${var.name}"
  availability_zones = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id             = "${var.vpc_id}"
  cidr_block         = "${local.private_cidr_block}"
  type               = "private"

  # Map of AZ names to NAT Gateway IDs was created in "public_subnets" module. Assign it to `az_ngw_ids` input
  az_ngw_ids         = "${module.public_subnets.az_ngw_ids}"
}
```

gives an error

> module.private_subnets.aws_route.default: aws_route.default: value of 'count' cannot be computed

because of this line
```hcl
az_ngw_ids         = "${module.public_subnets.az_ngw_ids}"
```

looks like TF can't calculate dynamic `count` using a Map variable from __another__ module .
(I have not found a solution yet).

On the other hand, this code works OK:

```hcl
locals {
  public_cidr_block  = "${cidrsubnet(var.vpc_cidr, 1, 0)}"
  private_cidr_block = "${cidrsubnet(var.vpc_cidr, 1, 1)}"
}

module "public_subnets" {
  source              = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=init"
  namespace           = "${var.namespace}"
  stage               = "${var.stage}"
  name                = "${var.name}"
  availability_zones  = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id              = "${var.vpc_id}"
  cidr_block          = "${local.public_cidr_block}"
  type                = "public"
  igw_id              = "${var.igw_id}"
  nat_gateway_enabled = "true"
}

module "private_subnets" {
  source             = "git::https://github.com/cloudposse/terraform-aws-multi-az-subnets.git?ref=init"
   namespace          = "${var.namespace}"
  stage              = "${var.stage}"
  name               = "${var.name}"
  availability_zones = ["us-east-1a", "us-east-1b", "us-east-1c"]
  vpc_id             = "${var.vpc_id}"
  cidr_block         = "${local.private_cidr_block}"
  type               = "private"

  # Map of AZ names to NAT Gateway IDs was created in "public_subnets" module. Assign it to `az_ngw_ids` input
  az_ngw_ids = {
    us-east-1a = "ngw-3213232"
    us-east-1b = "ngw-4234343"
    us-east-1c = "ngw-8676764"
  }
}
```

but it's not very useful because we need to know the NAT Gateways IDs since we are not getting them from the public subnets.

@osterman @comeanother @SweetOps 
What are your thoughts on all the above?
Should we try to find a fix for `terraform-aws-multi-az-subnets` invocation?
Should we just use `terraform-aws-named-subnets` and maybe improve it to cover more use-cases?
Should we improve `terraform-aws-dynamic-subnets` to cover more use-cases?

**Comment by comeanother** [comment]
@aknysh I think, we should use `terraform-aws-named-subnets` as a basis for further subnet modules.
And extend its capabilities.

I think, that `terraform-aws-dynamic-subnets` and `terraform-aws-multi-az-subnets` have too narrow range of use cases. As we're limited by one subnet per AZ.

If we add `availability_zones` list capability to `terraform-aws-named-subnets`, we will get one general module for subnets.

Imagine, if we're using `terraform-aws-dynamic-subnets` or `terraform-aws-multi-az-subnets` and we would need to add one (or a buch) more subnet to a specific AZ.
We would need to select another CIDR for that and perform additional calculation, when the first CIDR is not fully utilized.

**Comment by osterman** [comment]
I think the calculus of supporting named subnets + multi-az + nat gateways is way to complicated. The `mulit-az` subnet module is ideal for embedding in modules. For example, rds/redis/beanstalk/etc can define their own subnets with this. 

`dynamic-subnets` optmizes for a simple flat network with public and private subnets. 
`named-subnets` optimizes for lots of subnets in a single AZ (less-common use-case). 
`multi-az-subnets` optimizes for a single subnet in many AZs (very common use-case).

File: variables.tf [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+variable "availability_zones" {
+  type        = "list"
+  default     = []
+  description = "List of Availability Zones (e.g. `['us-east-1a', 'us-east-1b', 'us-east-1c']`)"
+}
+
+variable "max_subnets" {
+  default     = "16"
```
Comment (osterman) [inline variables.tf]: This value is too high. It implies there couple be 16 AZs, when I think the max is 6.

**Pull request merged**